src/libexpr-c/nix_api_expr.h

```diff
@@ -129,7 +129,8 @@ nix_err nix_value_call_multi(
  *
  * This function converts these Values into their final type.
  *
- * @note This function is mainly needed before calling @ref getters, but not for API calls that return a `Value`.
+ * @note You do not need this function as long as you only use API calls that return a `Value`. This function is mainly
+ * needed before calling @ref getters.
```

Comment:
```suggestion
 * @note This function is should be called before calling `nix_get_type()` (and therefore before most other @ref getters), because lazy evaluation should be transparent to users.
```

This sets callers up for https://github.com/NixOS/nix/issues/10499 and you should generally want to write small robust functions anyway, not micro-optimizing away a handful of force calls, where each such removal might be a mistake.
I don't want to turn that into docs though, because the C API does require that we e.g. force and check the type before getting e.g. the int or whatever is inside. Those things are more easily conveyed by reading individual function docs than trying to make some vague generalizations.
The part `(and therefore before most other @ref getters)` probably belongs in `nix_get_type()` actually.

Reply:
i thought, "you need to force first if and only if the api function does not return a `Value`" was a beautifully simple rule. is it wrong?